### ISSUE TITLE
Why is running a cluster (2 servers) slower than one

### ISSUE THREAD

**Issue by amamh** [dotnet/orleans#1336]
# Setup

I have two different PC's, and I have 2 servers and 1 client running:

- First PC:
  - First server
- Second PC:
  - Second server
  - A client

I'm using SQL SystemStore:

    <SystemStore SystemStoreType ="SqlServer" 
              DeploymentId="OrleansTest" 
              DataConnectionString="Data Source=.\sqlexpress;Initial Catalog=Orleans;Integrated Security=True;Pooling=False;Max Pool Size=200;Asynchronous Processing=True;MultipleActiveResultSets=True" AdoInvariant="System.Data.SqlClient" />

(On the second machine, I change the connection string to make it work)

I can see everything reported correctly in my SQL database. Table `OrleansMembershipTable` shows both servers as it should.

I have the following loop in the client:

            foreach (var instrument in _instrumentsByRic)
            {
                    var grain = GrainClient.GrainFactory.GetGrain<IIndexGrain>(instrument.RIC);
                    await grain.SetInstrument(instrument);
            }

Basically, this triggers **activation of about 20k grains** and sends some data `instrument` to them.

# Observation

That loop in the client takes an extremely long time (I waited for more than 10 minutes). However, **if I disable the second server (on the second machine) the loop then finishes in about 35 seconds**.

# Question

**Is grain activation supposed to be so slow when running a cluster compared to running only one server?**

I thought having a cluster should make things faster not slower, maybe it's slower for the client because it has to manage talking to two different servers?

---------------
Please let me know if there is anything that's unclear.

**Comment by jason-bragg** [comment]
In general, I'm aware of no reason for a cluster to slow down that significantly from simply adding a silo.

Having said that, the loop you are using awaits each call, which essentially serializes the operations.  This makes it rather pointless to add more silos.  I suggest adding the tasks to a list, then awaiting them all, this will allow for parallelization of the operations, which is necessary to gain performance benefits from distributed processing.

Also, do you have a gateway setup per silo?  If you've only one gateway, all your calls will be routed through that single gateway, meaning any work distributed to the second silo will incur additional (unnecessary) hops.

**Comment by amamh** [comment]
Great tip! Should have thought of that. I'll try it and report back.

I'm using one gateway per silo.

**Comment by jason-bragg** [comment]
As a general target, for a scenario like you've described, unless the instrument data is a large complex object, I'd expect to see about ~7k grain calls per second, so 20k requests should take ~3 seconds.

**Comment by amamh** [comment]
It does indeed increase performance a lot as expected, it takes 1/4 of the time before on my 4-cores i5. That's one server, not a cluster.

I can't test with two servers at the moment unfortunately, but I'll report back once I'm able to.

Out of interest, why do you say that you expect 7k grain calls per second? wouldn't that depend on how fast the hardware is and how complex the object (`instrument` in this example) is?

My `instrument` object has about 20 `string` fields, I'm not sure if that qualifies as complex or not.

**Comment by sergeybykov** [comment]
Going from 1 server to 2, you may see limited gains or no gains at all in some cases, depending on the cost of grain initialization. In that transition, you effectively turn 50% of internal runtime call, such as registration of grains in the directory and placement/forwarding, from being local, with only deep copying of only mutable arguments and no serialization, to remote messages with full serialization. The additional cost of remote messaging may negate a good chunk of the added capacity. 

This is kind of a corner case because when you add a third silo you'll only increase the %-age of remote messages from 50% to 66%, which makes much less of a difference, and even less so for more silos.

**Comment by amamh** [comment]
Thanks for the reply. I didn't test with a local severer, in the single
server case it was still a remote server. Also, my object `instrument` is
immutable so it shouldn't be cloned I believe.

**Comment by sergeybykov** [comment]
It's not about the server being local, it's rather about its silo components talking to other silo components within the same process and memory space, as opposed to another silo process on another server. Likewise, most of it is about arguments, and their immutability, of internal intra-silo/cross-silo system calls.

As for clients, we do strongly recommend to perform any measurements when each silo and each client process run on different machines in order to exclude performance interference between them.

**Comment by amamh** [comment]
Oh sorry, I see what you are saying now, I misunderstood.

I think that makes sense to me, it's hard to judge without knowing the internals of Orleans. I thought maybe the client will simply send 50% of the activation calls to the local server and 50% to the remote one, so it would save some time (sending some to local instead of remote, so no serialisation). Based on what you said, though, it seems like that's not how it works.

**Comment by sergeybykov** [comment]
Yes, network proximity is not that important. The cost of messaging - very much so.

**Comment by jason-bragg** [comment]
>why do you say that you expect 7k grain calls per second? wouldn't that depend on how fast the hardware is and how complex the object ( instrument  in this example) is?

Performance is, as you've eluded, dependent on many factors.  My 7k grain calls per second was a very rough [swag](https://en.wikipedia.org/wiki/Scientific_Wild-Ass_Guess) based off of what we've seen in our load tests.  In our distributed environment load tests we see roughly ~15k grain calls per second, but these calls use small objects and do no meaningful work.  These tests mostly measure the cost of messaging within a cluster.  If the grains do real work or the calls contain more complex objects, these numbers fall to somewhere between 7-9k grain calls per second.

Given that I've no idea what your processing does, how complex your calls are, or what hardware you're running on, I through out a conservative estimate, to help you gauge whether you're even in the ball park regarding performance expectations.

As Sergey elucidated, since you're sending only a single call to each grain, each call incurs an initial grain activation, which adds cost beyond the messaging inherent in the grain call itself, so the ~7k is probably a bit high.

**Comment by gabikliot** [comment]
I will close that for now. Feel free to re-open if you have more questions.
